### PR TITLE
CloudWatch Logs: Queries in an expression should run synchronously

### LINKED ISSUE #63735: CloudWatch Logs: applying Reduce expression results in an error
**What happened**:
Originally reported [here](https://github.com/grafana/grafana/issues/63388):
>I have not been able to apply panel expressions (like Reduce) to CloudWatch Logs queries either: I always get an error.

**How to reproduce it (as minimally and precisely as possible)**:
1. Set up a CloudWatch Logs query
2. Add an expression and apply Reduce
Something like this:
Screenshot 2023-02-24 at 15 54 37

Result:
An error in the plugin has occurred

**Anything else we need to know?**:
Debugging has shown:
We observe that the expression causes it to go through a time series query [here](https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudwatch/cloudwatch.go#L182)
and the error that it hits is [this one](https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudwatch/models/cloudwatch_query.go#L295)

**Environment**:
- Grafana version: ~9.4.0
- Data source type & version: CloudWatch Logs

### PR DESCRIPTION
**Pull request by iwysiu**
Cloudwatch Logs queries are generally run asynchronously in parts that are managed by the frontend (the frontend sends separate requests to start the query and then to poll for results). However, when a expression is used, it expects the query results to be returned from the first request. This pr adds a flag to the expression query that signals to the cloudwatch datasource that it should run the entire query synchronously in the backend and return the results (which we already handle for alerting).

The fixed bug can be reproed in the external dev as below:
Screenshot 2023-03-09 at 2 29 38 PM

**Which issue(s) does this PR fix?**:

Fixes #63735 

**Special notes for your reviewer**:
Note that this doesn't solve the separate "wide format" multiple time series issue https://github.com/grafana/grafana/issues/63388, that will still require changing how we create data frames.
Also I suspect that this may be an issue in async redshift/athena.

### CODE REVIEW

File: pkg/tsdb/cloudwatch/log_alert.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ const (
 	alertPollPeriod  = time.Second
 )
 
-func (e *cloudWatchExecutor) executeLogAlertQuery(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
```
Comment (sarahzinger) [inline pkg/tsdb/cloudwatch/log_alert.go:21]: Now that this is used for things beyond alerting should it go in a separate file, seems a bit weird to have it in a file called log_alert probably right? 

Also sort of basic on my part but how would one know if they should use a Sync vs Async Query?

File: pkg/expr/nodes.go [hunk 2]
```diff
@@ -227,6 +229,7 @@ func (dn *DSNode) Execute(ctx context.Context, now time.Time, _ mathexp.Vars, s
 		},
 		Headers: dn.request.Headers,
 	}
+	req.Headers[FromExpressionHeaderName] = "true"
```
Comment (sarahzinger) [inline pkg/expr/nodes.go:232]: how do other datasources deal with this problem? It feels worth asking grafana plugins platform folks

**Review by sarahzinger — COMMENTED** [review]
Cool catch and makes sense to me! Might be good to get someone from plugins platform to weigh-in maybe? It feels like this an interesting edge case that could be at least better documented as part of our api or something? Otherwise I'm not sure how any other plugin dev would know to do this sort of thing you know?

Also lets add tests!

File: pkg/expr/nodes.go [hunk 3]
```diff
@@ -21,6 +21,8 @@ var (
 	logger = log.New("expr")
 )
 
+const FromExpressionHeaderName = "FromExpression"
```
Comment (kylebrandt) [inline pkg/expr/nodes.go:24]: Since this is a header, we generally do `X-Grafana` , perhaps `X-Grafana-From-SSE`

Comment (kylebrandt) [reply pkg/expr/nodes.go:24]: Oh, I'm not sure these become HTTP headers, and I guess it matches FromAlert, so okay to leave it.

**Review by kylebrandt — APPROVED** [review]

**Review by iwysiu — COMMENTED** [review]
> Cool catch and makes sense to me! Might be good to get someone from plugins platform to weigh-in maybe? It feels like this an interesting edge case that could be at least better documented as part of our api or something? Otherwise I'm not sure how any other plugin dev would know to do this sort of thing you know?

It would probably make sense to document this somewhere. I actually don't think that other people are doing async datasources, at least right now, but other people might. Also, tests added! I needed to make one of the functions mock-able to do it.

**Review by kevinwcyu — APPROVED** [review]

**Review by sarahzinger — APPROVED** [review]
🚀

**Pull request merged**